### PR TITLE
🖍 Added shopping tag flip on too far position to the right

### LINKED ISSUE #37028: [amp story shopping] Shopping tag positioning on edge of page
### Description

If a shopping tag is too far to the right of the page 
- flip it so the dot is on the right
- transform element so that indicator dot is in the same position as it was in before the flip.

Current:
Screen Shot 2021-11-22 at 12 39 34 PM

Expected:
Screen Shot 2021-11-22 at 12 41 09 PM

Ensure this works with RTL.

### PR DESCRIPTION
**Pull request by jshamble**
Added the ability to have a shopping tag auto-change it's pill position to the left of the dot if it is too far to the right of the screen.

Also flips the dot to align with the correct spot for RTL languages.

Fixes #37028

Normal LTR           |  Flipped LTR
:-------------------------:|:-------------------------:
Screen Shot 2021-12-13 at 8 24 40 PM |  Screen Shot 2021-12-13 at 8 32 00 PM

Normal RTL           |  Flipped RTL
:-------------------------:|:-------------------------:
Screen Shot 2022-01-06 at 10 01 38 PM | Screen Shot 2022-01-06 at 10 01 56 PM

### CODE REVIEW

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (coreymasanto) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:95]: Does this calculation need to also take into account the width of the tag in order to determine the distance?

**Review by coreymasanto — COMMENTED** [review]
Could you add a screenshot of a story that includes both normal and flipped shopping tags?

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:95]: What Corey said, + RTL support :))
Feel free to look into what can be re-used from existing tooltips. Also for anything accessibility. Have we considered building on top of the existing tooltips to get all of this for free?

Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:95]: Will add RTL support once PM discussion and approval on currency format is complete. See my comment on https://github.com/ampproject/amphtml/pull/37187#discussion_r768276620

Comment (gmajoulet) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:95]: This looks unrelated to the currency string. The shopping tag should always be positioned correctly for both RTL and LTR.

Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:95]: I will take a look into RTL for this ticket after the specs ticket is merged #37186

Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: RTL Support was added in with this PR: #37249

Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: also added in the width of the tag calculation

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:178]: To recap our explorations:
It looks like RTL state is set when the `body` or `html` element has the `dir` attribute.
A separate RTL example document will be necessary to test RTL that has `dir="rtl"` on `body` or `html`.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:172]: These will trigger relayouts and is precisely what has to be in a `measureElement`. It forces the browser to recalculate all elements positions after it's already done it so it can give you accurate measures.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:178]: Please get the RTL_STATE from the store service, it should be the one source of truth over the DOM.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:179]: You're repeating this magic number twice, please add a top level file constant

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:180]: Should this have the `i-amphtml-` prefix?

**Comment by processprocess** [comment]
There's another piece to this concept that needs to be worked into this PR.
The element needs to be transformed so that the indicator dot is in the same place as it was before the flip. This ensures that the tag will be on screen.

Screen Shot 2022-01-13 at 10 15 01 AM

An approach:
If the element is off-screen on the right:
translateX by -100% + the size of the dot.
If off-screen on the left, a flipped version of the same equation can be used.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.css
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.css:38]: row is implied by default, this might not be necessary.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:184]: Instead of adding and removing, we could toggle an `amp-story-shopping-tag-inner-flipped` class that overrides the flex direction. That will simplify this logic.

**Comment by processprocess** [comment]
We need to hide the button until the calculation is made and the element is flipped. Otherwise we see it flash on one side then flip to the other:

This could be accomplished by setting opacity to 0 initially and setting it to 1 after the check and positioning is done.

Also the margin between the button and pill is gone. 
Please bring that back.

https://user-images.githubusercontent.com/3860311/149560631-f2a38d08-0d24-44ca-aedc-31db65e6ee50.mp4

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:192]: got it, thanks for the clarification

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: Even better, found a way to do it without magic numbers.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: The class in the .css file is .amp-story-shopping-tag-inner, so I'm not sure I need the prefix.

Comment (gmajoulet) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: Why doesn't `amp-story-shopping-tag-inner` have the prefix? They all should have it

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:82]: For AMP components (`amp-story`, `amp-story-shopping-tag`, etc), please get the cached measurements using `el.getLayoutBox()` instead of re-measuring

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:87]: Please get `StateProperty.PAGE_SIZE` from the store service

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:90]: You're changing a class in the DOM, which updates the DOM/layout, this needs to be in `mutateElement`

1. Measuring something that might trigger a reflow: measure
2. Writing in the DOM: mutate

Please spend some time looking into the resources I shared with you (or ping me if you want them again / if they're not clear and you want to talk about it). This is really important web perf and frame management stuff.

Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:90]: Soudns good, but is it valid to call mutatelement within a measure element?

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:82]: I am getting this when I try to use `. getLayoutBox()` , are you sure you want me to use it?

Forbidden: ".getLayoutBox". getLayoutSize/Box APIs are being deprecated. Please contact the @ampproject/wg-performance for questions.eslint(local/no-forbidden-terms)

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: I added prefixes to the css classes and they will be pushed in the next commit.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:87]: We are not using the width of the Page Size, we are using the width of the story page. The page size will return the width and height of the whole page (including the big border around the mobile rectangle in the middle), so it will not work for desktop if we get StateProperty.PAGE_SIZE from the store service; @processprocess any thoughts? I can't think of a way to calculate that number just using the page size width alone from the store service without involving some sort of DOM measurement.

Comment (processprocess) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:87]: I can confirm that `PAGE_SIZE` is the size of the window and not the page.

[SET_PAGE_SIZE](https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/amp-story.js#L1497) is currently updated with the [window dimensions](https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/amp-story.js#L355), which is unexpected in the desktop experience.

I think it would be worthwhile to fix this in a separate PR.
If we were to do that we would also need to update the CSS overrides [here](https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/amp-story-desktop-one-panel.css#L23-L24).

@gmajoulet Do you think that fix should block this PR? I'm leaning that we should go ahead and fix this now so we can use it in the future.

Comment (gmajoulet) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: This is 100% a bug, introduced here: https://github.com/ampproject/amphtml/commit/3012bf62aa194478825d34f04b909d6c4cadd061#diff-94f9e2fa2969f98d22c37d9e1f195adbe79ebda93b4ffedf4eab34327238418fL570

Yes, it should be fixed before we move forward with this PR, that should use `PAGE_SIZE`

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (samouri) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: If ok to make async, you can make the measure nonblocking via using either InOb or ResOb:

https://github.com/ampproject/amphtml/blob/a993fd73d200a544029296d9e40e66d33c6b5718/src/core/dom/layout/intersection-no-root.js#L69-L81

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: This is bug fixed with #37435

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js]: We are using offsetLeft and offsetWidth instead of getLayoutBox() because
the correct measurements are not taken into account when using a CSS transform (such as translate),
which we are using in the i-amphtml-amp-story-shopping-tag-inner-flipped class.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (coreymasanto) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:251]: This method is being called in response to both the `SHOPPING_DATA` and `PAGE_SIZE` actions. Is this necessary?

File: examples/visual-tests/amp-story/amp-story-shopping-rtl.html
Comment (jshamble) [reply examples/visual-tests/amp-story/amp-story-shopping-rtl.html:1]: It is necessary - rtl is only checked on the top level tags - so we had to create another file to test it properly rather than modifying it to use trl on the subtrees. (we are assuming that user will not require a 'mixed' case , where one page could be RTL and another is LTR)

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:251]: Good point - needed to add `true /** callToInitialize */` to get it to work.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:98]: That is how it was originally. I was keeping it there to avoid doing unnecessary calculations in the mutate element - I'll make your change, but if anyone has any objections feel free to re-open and comment.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:74]: Good point, however I tried this and since it was in a measure element, it caused a bit of problems with the DOM syncing up properly as I would be returning a promise, not a boolean. I think it would be better to just rename the method to something like `flipTagIfOffscreen_`.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (coreymasanto) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:117]: Since this call is independent of tag flip logic, I think you can move it out of this method and place it right after the `this.flipTagIfOffscreen_()` call in the PAGE_SIZE callback

**Review by coreymasanto — APPROVED** [review]

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:117]: I tired that, and even after adding in a mutate element, it will have problems syncing the flip state during initialization in the browser tests I performed (there is a small window where the tag is not flipped at the beginning, and we are hiding that by setting the opacity to zero in a CSS class). Having this here ensures that the tag will not be shown in that state.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:102]: plz get the `mutateElement` outside of the `measureElement`. You can do:

```
let shouldFlip;

this.measureMutateElement(() => {
  shouldFlip = bool;
}, () => {
  toggle(shouldFlip);
});
```

File: extensions/amp-story/1.0/amp-story.js
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js:917]: This should not be needed in this PR?

Comment (jshamble) [reply extensions/amp-story/1.0/amp-story.js:917]: Upon further testing it doesn't work on initialization unless I move the call here -

Comment (gmajoulet) [reply extensions/amp-story/1.0/amp-story.js:917]: Yes please! This could have a lot of CLS implications, we need to look into it carefully cc @mszylkowski

Comment (jshamble) [reply extensions/amp-story/1.0/amp-story.js:917]: I have opened up #37520 and added you and @mszylkowski as a reviewer, I'll merge this PR after that one is in.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js
Comment (gmajoulet) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js:106]: Re: this is not doing any DOM reading operation so it doesn't need to be in a measureElement context. Let me know if I can help you with more documentation for these

**Review by gmajoulet — APPROVED** [review]

**Review by processprocess — APPROVED** [review]
Looks awesome!

**Comment by processprocess** [comment]
The margins are broken in RTL. Please take a look into this:

Screen Shot 2022-02-02 at 1 47 52 PM

**Pull request merged**
